### PR TITLE
Kjezek/rlp decoder

### PR DESCRIPTION
**Pull request by kjezek**
This PR implements RLP decoder.  It interprets the input `[]byte` stream and returns an` rlp.Item`

### CODE REVIEW

File: go/database/mpt/rlp/rlp.go [hunk 1]
```diff
@@ -50,6 +51,90 @@ func EncodeInto(dst []byte, item Item) []byte {
...
+	item, consumed, err := decode(rlp)
+	if consumed < uint64(len(rlp)) {
+		return nil, fmt.Errorf("unexpected trailing data: %x", rlp[consumed:])
+	}
+	return item, err
+}
+
+// decode decodes an RLP stream into an item.
+// It checks first byte of the RLP stream to determine the type of the item.
+// Based on the type, it decodes the type.
```
Comment (HerbertJordan) [inline go/database/mpt/rlp/rlp.go:65]: nit: implementation details should go into the function body

File: go/database/mpt/rlp/rlp.go [hunk 2]
```diff
@@ -253,3 +364,45 @@ func (i BigInt) getEncodedLength() int {
...
 }
+
+// readNumber interprets the first 'slen' bytes of 'b' as a big-endian integer.
+// It returns the resulting integer and any error encountered.
+//
+// The function expects 'slen' to be the size of the integer to be read from 'b'.
+// If 'slen' is greater than the length of 'b', an error is returned.
+//
+// The function supports reading integers of up to 8 bytes (64 bits).
+// The integer is constructed by shifting and OR-ing the bytes from 'b'.
```
Comment (HerbertJordan) [inline go/database/mpt/rlp/rlp.go:375]: nit: this line is an implementation detail

File: go/database/mpt/rlp/rlp.go [hunk 3]
```diff
@@ -253,3 +364,45 @@ func (i BigInt) getEncodedLength() int {
...
+		s = uint64(b[0])<<32 | uint64(b[1])<<24 | uint64(b[2])<<16 | uint64(b[3])<<8 | uint64(b[4])
+	case 6:
+		s = uint64(b[0])<<40 | uint64(b[1])<<32 | uint64(b[2])<<24 | uint64(b[3])<<16 | uint64(b[4])<<8 | uint64(b[5])
+	case 7:
+		s = uint64(b[0])<<48 | uint64(b[1])<<40 | uint64(b[2])<<32 | uint64(b[3])<<24 | uint64(b[4])<<16 | uint64(b[5])<<8 | uint64(b[6])
+	case 8:
+		s = uint64(b[0])<<56 | uint64(b[1])<<48 | uint64(b[2])<<40 | uint64(b[3])<<32 | uint64(b[4])<<24 | uint64(b[5])<<16 | uint64(b[6])<<8 | uint64(b[7])
+	}
+
+	return s, nil
```
Comment (HerbertJordan) [inline go/database/mpt/rlp/rlp.go:407]: Maybe shorten this to
```suggestion
	var s uint64
	for i := byte(0); i < slen; i++ {
		s = s<<8 | uint64(b[i])
	}
	return s, nil
```

File: go/database/mpt/rlp/rlp_test.go [hunk 4]
```diff
@@ -12,82 +12,270 @@ package rlp
...
 	"bytes"
+	"fmt"
 	"math/big"
 	"testing"
 
 	"github.com/Fantom-foundation/Carmen/go/common"
 )
 
 func TestEncoding_EncodeStrings(t *testing.T) {
+	testWithRlpStrings(t, func(t *testing.T, rlp []byte, item String) {
```
Comment (HerbertJordan) [inline go/database/mpt/rlp/rlp_test.go:23]: nit: rename "testWithRlpString" => "forEachRlpStringTest" ? Same for the others below.

**Review by HerbertJordan — APPROVED** [review]
Looks good, some minor cosmetic comments.

File: go/database/mpt/rlp/rlp.go (hunk 1 above)
Comment (kjezek) [reply go/database/mpt/rlp/rlp.go:65]: right, moved in the body

File: go/database/mpt/rlp/rlp.go (hunk 2 above)
Comment (kjezek) [reply go/database/mpt/rlp/rlp.go:375]: right, removed

File: go/database/mpt/rlp/rlp_test.go (hunk 4 above)
Comment (kjezek) [reply go/database/mpt/rlp/rlp_test.go:23]: ok, renamed

**Review by HerbertJordan — APPROVED** [review]
